tests/standalone_display_test.cpp

```diff
@@ -0,0 +1,217 @@
+  }
+  return 0;
+}
+
+void init_display(android::HwcDisplay* hwc_display, hwc2_layer_t& layer_id) {
+  int32_t width = 0;
+  int32_t height = 0;
+  hwc2_config_t config = 0;
+  hwc_display->GetActiveConfig(&config);
+  hwc_display->GetDisplayAttribute(config,
+                                   static_cast<int32_t>(HWC2::Attribute::Width),
+                                   &width);
+  hwc_display->GetDisplayAttribute(config,
+                                   static_cast<int32_t>(
+                                       HWC2::Attribute::Height),
+                                   &height);
+
+  for (uint8_t i = 0; i < fb_num; i++) {
+    g_bo[i].width = width;
+    g_bo[i].height = height;
+    modeset_create_fb(hwc_display->GetPipe().device->GetFd(), &g_bo[i], i);
+  }
+  /// create and set layer info
+  hwc_display->CreateLayer(&layer_id);
+  android::HwcLayer* layer = hwc_display->get_layer(layer_id);
+  layer->SetLayerBlendMode((int32_t)HWC2::BlendMode::Premultiplied);
+  layer->SetValidatedType(HWC2::Composition::Device);
+  layer->AcceptTypeChange();
+  hwc_rect_t display_frame = {0, 0, width, height};
+  layer->SetLayerDisplayFrame(display_frame);
+  hwc_frect_t source_crop = {0.0, 0.0, static_cast<float>(width),
+                             static_cast<float>(height)};
+  layer->SetLayerSourceCrop(source_crop);
+}
+
+int modeset_create_fb(int fd, Buffer_object* bo, uint8_t index) {
+  int ret;
+  /// 1 create dumb buffer
+  struct drm_mode_create_dumb create = {};
+  memset(&create, 0, sizeof(create));
+  create.width = bo->width;
+  create.height = bo->height;
+  create.bpp = 32;
+  create.handle = 0;
+  ret = drmIoctl(fd, DRM_IOCTL_MODE_CREATE_DUMB, &create);
+  if (ret < 0) {
+    fprintf(stderr, "cannot create dumb buffer (%d): %m\n", errno);
+    return -errno;
+  }
+  bo->stride = create.pitch;
+  bo->size = create.size;
+  bo->handle = create.handle;
+
+  // 2 ADDFB2
+  struct drm_mode_fb_cmd2 f;
+  memset(&f, 0, sizeof(f));
+  f.width = bo->width;
+  f.height = bo->height;
+  f.pixel_format = DRM_FORMAT_ARGB8888;
+  f.flags = 0;
+  f.handles[0] = bo->handle;
+  f.modifier[0] = 0;
+  f.pitches[0] = bo->stride;
+  f.offsets[0] = 0;
+
+  ret = drmIoctl(fd, DRM_IOCTL_MODE_ADDFB2, &f);
+  if (ret < 0) {
+    fprintf(stderr, "cannot addfb2 buffer (%d): %m\n", errno);
+    return -errno;
+  }
+  bo->fb_id = f.fb_id;
+  /// 3 map
+  struct drm_mode_map_dumb map = {};
+  memset(&map, 0, sizeof(map));
+  map.handle = bo->handle;
+  ret = drmIoctl(fd, DRM_IOCTL_MODE_MAP_DUMB, &map);
+  if (ret) {
+    fprintf(stderr, "cannot map dumb buffer (%d): %m\n", errno);
+    ret = -errno;
+  }
+
+  bo->vaddr = (uint8_t*)mmap(0, create.size, PROT_READ | PROT_WRITE, MAP_SHARED,
+                             fd, map.offset);
+  if (bo->vaddr == MAP_FAILED) {
+    fprintf(stderr, "cannot mmap dumb buffer (%d): %m\n", errno);
+    ret = -errno;
+  }
+  /// 4 fill color info
+  fill_buffer(bo, index);
+  /// 5 unmap
+  munmap(bo->vaddr, bo->size);
```

Comment:
mmap and munmap need move into fill_buffer function. Also need let main directly call fill_buffer. fill_buffer should not be called in create_fb.